core/src/main/scala/org/typelevel/vault/Locker.scala

```diff
@@ -39,15 +39,31 @@ final class Locker private (private val unique: Unique.Token, private val a: Any
    * @param k
    *   The key to check, if the internal Unique value matches then this Locker can be unlocked as the specifed value
    */
-  def unlock[A](k: Key[A]): Option[A] = Locker.unlock(k, this)
+  def unlock[A](k: LookupKey[A]): Option[A] = Locker.unlock(k, this)
+
+  /**
+   * Retrieve the value from the Locker. If the reference equality
+   * instance backed by a `Unique` value is the same then allows
+   * conversion to that type, otherwise as it does not match
+   * then this will be `None`
+   *
+   * @param k The key to check, if the internal Unique value matches
+   * then this Locker can be unlocked as the specifed value
+   */
+  private[vault] def unlock[A](k: Key[A]): Option[A] = unlock(k: LookupKey[A])
 }
 
 object Locker {
 
   /**
    * Put a single value into a Locker
    */
-  def lock[A](k: Key[A], a: A): Locker = new Locker(k.unique, a.asInstanceOf[Any])
+  def lock[A](k: InsertKey[A], a: A): Locker = new Locker(k.unique, k.in(a))
+
+  /**
+   * Put a single value into a Locker
+   */
+  def lock[A](k: Key[A], a: A): Locker = lock(k: InsertKey[A], a)
```

Comment:
Overloads make me queasy and overloads where one is a subtype of another make me queasier still.  I guess this is necessary for bincompat?  Is there a different name we could give the `InsertKey` version?

Reply:
Or could we have just made this `private[vault]`?

Reply:
We have the same basic issue with `unlock`.

Reply:
This is needed for bincompat, and no `private[vault]` does not work because it disappears the static forwarders. I ran into this same issue in https://github.com/typelevel/fs2/pull/2466.

OTOH, there is precedence for not caring about static forwarders in https://github.com/http4s/http4s/pull/5071#issuecomment-898705809.

Reply:
I'm not sure that was a good precedent.  And it seems we're stuck with the overload on `delete` anyway.  I find it a little confusing, but I don't see a good way out.

Reply:
Maybe an unpopular opinion, one way out would be to deprecate the vault companion object based methods (I always found the duplication confusing, does anyone actually use them like this?) in favor of just calling these methods directly on the vault instance, and then just not add these new methods to the companion object.

Reply:
I didn't realize they existed on the companions.  I think that's a good idea.  The only weird one then is the overloaded `Lock.lock`, right? And perhaps the new one could be called `apply`?

Reply:
Cool, let me give that a try 👍